File: modules/platforms/cpp/core-test/src/date_time_test.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <boost/test/unit_test.hpp>

#include <ignite/time.h>
#include <ignite/date.h>
#include <ignite/timestamp.h>

#include <ignite/common/utils.h>

#include <ignite/test_utils.h>

using namespace ignite;
using namespace cache;
using namespace boost::unit_test;

/**
 * Check operators for type.
 * All args should refer to a different instances.
 * Also, val1 should be less then val2 and equeal val3.
 */
template<typename T>
void CheckOperators(const T& val1, const T& val2, const T& val3)
{
    BOOST_CHECK(&val1 != &val2);
    BOOST_CHECK(&val1 != &val3);
    BOOST_CHECK(&val2 != &val3);

    BOOST_CHECK(val1 == val1);
    BOOST_CHECK(val2 == val2);

    BOOST_CHECK(val1 == val3);
    BOOST_CHECK(val3 == val1);

    BOOST_CHECK(val1 != val2);
    BOOST_CHECK(val2 != val1);
    BOOST_CHECK(val3 != val2);
    BOOST_CHECK(val2 != val3);

    BOOST_CHECK(val1 < val2);
    BOOST_CHECK(val3 < val2);

    BOOST_CHECK(val1 <= val2);
    BOOST_CHECK(val3 <= val2);
    BOOST_CHECK(val1 <= val1);
    BOOST_CHECK(val2 <= val2);
    BOOST_CHECK(val1 <= val3);
    BOOST_CHECK(val3 <= val1);

    BOOST_CHECK(val2 > val1);
    BOOST_CHECK(val2 > val3);

    BOOST_CHECK(val2 >= val1);
    BOOST_CHECK(val2 >= val3);
    BOOST_CHECK(val1 >= val1);
    BOOST_CHECK(val2 >= val2);
    BOOST_CHECK(val1 >= val3);
    BOOST_CHECK(val3 >= val1);
}

void CheckTime(int hour, int mins, int sec)
{
    Time time = common::MakeTimeGmt(hour, mins, sec);
    tm res;

    common::TimeToCTm(time, res);

    BOOST_CHECK_EQUAL(res.tm_hour, hour);
    BOOST_CHECK_EQUAL(res.tm_min, mins);
    BOOST_CHECK_EQUAL(res.tm_sec, sec);
}

void CheckDate(int year, int mon, int day)
{
    Date date = common::MakeDateGmt(year, mon, day);
    tm res;

    common::DateToCTm(date, res);

    BOOST_CHECK_EQUAL(res.tm_year + 1900, year);
    BOOST_CHECK_EQUAL(res.tm_mon + 1, mon);
    BOOST_CHECK_EQUAL(res.tm_mday, day);
}

void CheckTimestamp(int year, int mon, int day, int hour, int mins, int sec, int ns)
{
    Timestamp ts = common::MakeTimestampGmt(year, mon, day, hour, mins, sec, ns);
    tm res;

    common::TimestampToCTm(ts, res);

    BOOST_CHECK_EQUAL(res.tm_year + 1900, year);
    BOOST_CHECK_EQUAL(res.tm_mon + 1, mon);
    BOOST_CHECK_EQUAL(res.tm_mday, day);
    BOOST_CHECK_EQUAL(res.tm_hour, hour);
    BOOST_CHECK_EQUAL(res.tm_min, mins);
    BOOST_CHECK_EQUAL(res.tm_sec, sec);
}

BOOST_AUTO_TEST_SUITE(DateTimeTestSuite)

BOOST_AUTO_TEST_CASE(TimeOperators1)
{
    Time val1(1);
    Time val2(2);
    Time val3(1);

    CheckOperators(val1, val2, val3);
}

BOOST_AUTO_TEST_CASE(TimeOperators2)
{
    Time val1(154362);
    Time val2(val1.GetMilliseconds() + 42363);
    Time val3(val1);

    CheckOperators(val1, val2, val3);
}

BOOST_AUTO_TEST_CASE(DateOperators1)
{
    Date val1(1);
    Date val2(2);
    Date val3(1);

    CheckOperators(val1, val2, val3);
}

BOOST_AUTO_TEST_CASE(DateOperators2)
{
    Date val1(154362);
    Date val2(val1.GetMilliseconds() + 42363);
    Date val3(val1);

    CheckOperators(val1, val2, val3);
}

BOOST_AUTO_TEST_CASE(TimestampOperators1)
{
    Timestamp val1(1);
    Timestamp val2(2);
    Timestamp val3(1);

    CheckOperators(val1, val2, val3);
}

BOOST_AUTO_TEST_CASE(TimestampOperators2)
{
    Timestamp val1(154362);
    Timestamp val2(val1.GetMilliseconds() + 42363);
    Timestamp val3(val1);

    CheckOperators(val1, val2, val3);
}

BOOST_AUTO_TEST_CASE(TimestampOperators3)
{
    Timestamp val1(42, 1);
    Timestamp val2(42, 2);
    Timestamp val3(42, 1);

    CheckOperators(val1, val2, val3);
}

BOOST_AUTO_TEST_CASE(TimestampOperators4)
{
    Timestamp val1(42, 154362);
    Timestamp val2(42, val1.GetSecondFraction() + 42363);
    Timestamp val3(42, val1.GetSecondFraction());

    CheckOperators(val1, val2, val3);
}

BOOST_AUTO_TEST_CASE(TimestampOperators5)
{
    Timestamp val1(154362, 154362);
    Timestamp val2(val1.GetMilliseconds() + 42363, val1.GetSecondFraction() + 42363);
    Timestamp val3(val1);

    CheckOperators(val1, val2, val3);
}

BOOST_AUTO_TEST_CASE(MakeTime)
{
    using namespace common;

    BOOST_CHECK_EQUAL(MakeTimeGmt(0, 0, 0).GetMilliseconds(), 0);
    BOOST_CHECK_EQUAL(MakeTimeGmt(23, 59, 59).GetMilliseconds(), 86399000);
    BOOST_CHECK_EQUAL(MakeTimeGmt(14, 23, 47).GetMilliseconds(), 51827000);
    BOOST_CHECK_EQUAL(MakeTimeGmt(0, 24, 12).GetMilliseconds(), 1452000);
    BOOST_CHECK_EQUAL(MakeTimeGmt(8, 0, 0).GetMilliseconds(), 28800000);

    BOOST_CHECK_EQUAL(MakeTimeGmt(0, 0, 0).GetSeconds(), 0);
    BOOST_CHECK_EQUAL(MakeTimeGmt(23, 59, 59).GetSeconds(), 86399);
    BOOST_CHECK_EQUAL(MakeTimeGmt(14, 23, 47).GetSeconds(), 51827);
    BOOST_CHECK_EQUAL(MakeTimeGmt(0, 24, 12).GetSeconds(), 1452);
    BOOST_CHECK_EQUAL(MakeTimeGmt(8, 0, 0).GetSeconds(), 28800);
}

BOOST_AUTO_TEST_CASE(MakeDate)
{
    using namespace common;

    BOOST_CHECK_EQUAL(MakeDateGmt(1970, 1, 1).GetMilliseconds(), 0);
    BOOST_CHECK_EQUAL(MakeDateGmt(2000, 12, 31).GetMilliseconds(), 978220800000);
    BOOST_CHECK_EQUAL(MakeDateGmt(2017, 3, 20).GetMilliseconds(), 1489968000000);

    BOOST_CHECK_EQUAL(MakeDateGmt(1970, 1, 1).GetSeconds(), 0);
    BOOST_CHECK_EQUAL(MakeDateGmt(2000, 12, 31).GetSeconds(), 978220800);
    BOOST_CHECK_EQUAL(MakeDateGmt(2017, 3, 20).GetSeconds(), 1489968000);
}

BOOST_AUTO_TEST_CASE(MakeTimestamp)
{
    using namespace common;

    BOOST_CHECK_EQUAL(MakeTimestampGmt(1970, 1, 1, 0, 0, 0, 0).GetMilliseconds(), 0);
    BOOST_CHECK_EQUAL(MakeTimestampGmt(2000, 12, 31, 23, 59, 59, 999999999).GetMilliseconds(), 978307199999);
    BOOST_CHECK_EQUAL(MakeTimestampGmt(2001, 9, 9, 1, 46, 39, 999999999).GetMilliseconds(), 999999999999);
    BOOST_CHECK_EQUAL(MakeTimestampGmt(2017, 3, 20, 18, 43, 19, 170038645).GetMilliseconds(), 1490035399170);

    BOOST_CHECK_EQUAL(MakeTimestampGmt(1970, 1, 1, 0, 0, 0, 0).GetSeconds(), 0);
    BOOST_CHECK_EQUAL(MakeTimestampGmt(2000, 12, 31, 23, 59, 59, 999999999).GetSeconds(), 978307199);
    BOOST_CHECK_EQUAL(MakeTimestampGmt(2001, 9, 9, 1, 46, 39, 999999999).GetSeconds(), 999999999);
    BOOST_CHECK_EQUAL(MakeTimestampGmt(2017, 3, 20, 18, 43, 19, 170038645).GetSeconds(), 1490035399);

    BOOST_CHECK_EQUAL(MakeTimestampGmt(1970, 1, 1, 0, 0, 0, 0).GetSecondFraction(), 0);
    BOOST_CHECK_EQUAL(MakeTimestampGmt(2000, 12, 31, 23, 59, 59, 999999999).GetSecondFraction(), 999999999);
    BOOST_CHECK_EQUAL(MakeTimestampGmt(2001, 9, 9, 1, 46, 39, 999999999).GetSecondFraction(), 999999999);
    BOOST_CHECK_EQUAL(MakeTimestampGmt(2017, 3, 20, 18, 43, 19, 170038645).GetSecondFraction(), 170038645);

    BOOST_CHECK_EQUAL(MakeTimestampGmt(1970, 1, 1, 0, 0, 0, 0).GetDate().GetMilliseconds(), 0);
    BOOST_CHECK_EQUAL(MakeTimestampGmt(2000, 12, 31, 23, 59, 59, 999999999).GetDate().GetMilliseconds(), 978307199999);
    BOOST_CHECK_EQUAL(MakeTimestampGmt(2001, 9, 9, 1, 46, 39, 999999999).GetDate().GetMilliseconds(), 999999999999);
    BOOST_CHECK_EQUAL(MakeTimestampGmt(2017, 3, 20, 18, 43, 19, 170038645).GetDate().GetMilliseconds(), 1490035399170);
}

BOOST_AUTO_TEST_CASE(CastTimeToTm)
{
    CheckTime(21, 8, 5);
    CheckTime(12, 41, 11);
    CheckTime(1, 28, 18);
    CheckTime(8, 12, 59);
    CheckTime(17, 52, 31);
    CheckTime(21, 56, 21);
}

BOOST_AUTO_TEST_CASE(CastDateToTm)
{
    CheckDate(2024, 8, 5);
    CheckDate(1987, 1, 1);
    CheckDate(1999, 2, 18);
    CheckDate(1997, 12, 9);
    CheckDate(2007, 12, 31);
    CheckDate(2001, 6, 21);
}

BOOST_AUTO_TEST_CASE(CastTimestampToTm)
{
    CheckTimestamp(1970, 1, 1, 0, 0, 0, 0);
    CheckTimestamp(2000, 12, 31, 23, 59, 59, 999999999);
    CheckTimestamp(2001, 9, 9, 1, 46, 39, 999999999);
    CheckTimestamp(2017, 3, 20, 18, 43, 19, 170038645);
    CheckTimestamp(2007, 12, 31, 19, 24, 44, 894375963);
    CheckTimestamp(2001, 6, 21, 13, 53, 2, 25346547);
}

BOOST_AUTO_TEST_SUITE_END()

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/DateTimeTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.ignite.internal.processors.query.calcite;

import java.sql.Time;
import java.sql.Timestamp;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.TimeZone;
import org.apache.ignite.Ignite;
import org.apache.ignite.IgniteCache;
import org.apache.ignite.cache.CacheMode;
import org.apache.ignite.cache.QueryEntity;
import org.apache.ignite.configuration.CacheConfiguration;
import org.apache.ignite.internal.IgniteEx;
import org.apache.ignite.internal.processors.query.QueryEngine;
import org.apache.ignite.internal.processors.query.calcite.util.Commons;
import org.apache.ignite.testframework.junits.WithSystemProperty;
import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
import org.junit.Test;

import static java.util.Collections.singletonList;

/** */
@WithSystemProperty(key = "calcite.debug", value = "true")
public class DateTimeTest extends GridCommonAbstractTest {
    /** */
    private static final SimpleDateFormat DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS");

    /** */
    private static QueryEngine queryEngine;

    /** {@inheritDoc} */
    @Override protected void beforeTestsStarted() throws Exception {
        Ignite grid = startGridsMultiThreaded(1);

        QueryEntity qryEnt = new QueryEntity();
        qryEnt.setKeyFieldName("ID");
        qryEnt.setKeyType(Integer.class.getName());
        qryEnt.setValueType(DateTimeEntry.class.getName());

        qryEnt.addQueryField("ID", Integer.class.getName(), null);
        qryEnt.addQueryField("JAVADATE", Date.class.getName(), null);
        qryEnt.addQueryField("SQLDATE", java.sql.Date.class.getName(), null);
        qryEnt.addQueryField("SQLTIME", Time.class.getName(), null);
        qryEnt.addQueryField("SQLTIMESTAMP", Timestamp.class.getName(), null);
        qryEnt.setTableName("datetimetable");

        final CacheConfiguration<Integer, DateTimeEntry> cfg = new CacheConfiguration<>(qryEnt.getTableName());

        cfg.setCacheMode(CacheMode.PARTITIONED)
            .setBackups(1)
            .setQueryEntities(singletonList(qryEnt))
            .setSqlSchema("PUBLIC");

        IgniteCache<Integer, DateTimeEntry> dateTimeCache = grid.createCache(cfg);

        dateTimeCache.put(1, new DateTimeEntry(1, javaDate("2020-10-01 12:00:00.000"),
            sqlDate("2020-10-01"), sqlTime("12:00:00"), sqlTimestamp("2020-10-01 12:00:00.000")));
        dateTimeCache.put(2, new DateTimeEntry(2, javaDate("2020-12-01 00:10:20.000"),
            sqlDate("2020-12-01"), sqlTime("00:10:20"), sqlTimestamp("2020-12-01 00:10:20.000")));
        dateTimeCache.put(3, new DateTimeEntry(3, javaDate("2020-10-20 13:15:00.000"),
            sqlDate("2020-10-20"), sqlTime("13:15:00"), sqlTimestamp("2020-10-20 13:15:00.000")));
        dateTimeCache.put(4, new DateTimeEntry(4, javaDate("2020-01-01 22:40:00.000"),
            sqlDate("2020-01-01"), sqlTime("22:40:00"), sqlTimestamp("2020-01-01 22:40:00.000")));

        queryEngine = Commons.lookupComponent(((IgniteEx)grid).context(), QueryEngine.class);

        awaitPartitionMapExchange();
    }

    /** */
    @Test
    public void testQuery1() throws Exception {
        checkQuery("SELECT SQLDATE FROM datetimetable where SQLTIME = '12:00:00'")
            .returns(sqlDate("2020-10-01"))
            .check();
    }

    /** */
    @Test
    public void testQuery2() throws Exception {
        checkQuery("SELECT SQLDATE FROM datetimetable where JAVADATE = ?")
            .withParams(javaDate("2020-12-01 00:10:20.000"))
            .returns(sqlDate("2020-12-01"))
            .check();
    }

    /** */
    @Test
    public void testQuery3() throws Exception {
        checkQuery("SELECT SQLDATE FROM datetimetable where JAVADATE = ?")
            .withParams(sqlTimestamp("2020-12-01 00:10:20.000"))
            .returns(sqlDate("2020-12-01"))
            .check();
    }

    /** */
    @Test
    public void testQuery4() throws Exception {
        checkQuery("SELECT MAX(SQLDATE) FROM datetimetable")
            .returns(sqlDate("2020-12-01"))
            .check();
    }

    /** */
    @Test
    public void testQuery5() throws Exception {
        checkQuery("SELECT MIN(SQLDATE) FROM datetimetable")
            .returns(sqlDate("2020-01-01"))
            .check();
    }

    /** */
    @Test
    public void testQuery6() throws Exception {
        checkQuery("SELECT JAVADATE FROM datetimetable WHERE SQLTIME = '13:15:00'")
            .returns(javaDate("2020-10-20 13:15:00.000"))
            .check();
    }

    /** */
    @Test
    public void testQuery7() throws Exception {
        checkQuery("SELECT t1.JAVADATE, t2.JAVADATE FROM datetimetable t1 " +
            "INNER JOIN " +
            "(SELECT JAVADATE, CAST(SQLTIMESTAMP AS TIME) AS CASTED_TIME FROM datetimetable) t2 " +
            "ON t1.SQLTIME = t2.CASTED_TIME " +
            "WHERE t2.JAVADATE = '2020-10-20 13:15:00.000'")
            .returns(javaDate("2020-10-20 13:15:00.000"), javaDate("2020-10-20 13:15:00.000"))
            .check();
    }

    /**
     * Test right date/time interpretation taking into account DST clock shift.
     */
    @Test
    public void testDstShift() throws Exception {
        TimeZone oldTz = TimeZone.getDefault();

        try {
            TimeZone.setDefault(TimeZone.getTimeZone("Europe/Moscow"));

            // Time zone change (EET->MSK) 1992-01-19 02:00:00 -> 1992-01-19 03:00:00
            checkQuery("select date '1992-01-19'").returns(sqlDate("1992-01-19")).check();
            checkQuery("select date '1992-01-18' + interval (1) days").returns(sqlDate("1992-01-19")).check();
            checkQuery("select date '1992-01-18' + interval (24) hours").returns(sqlDate("1992-01-19")).check();
            checkQuery("SELECT timestamp '1992-01-18 02:30:00' + interval (25) hours")
                .returns(sqlTimestamp("1992-01-19 03:30:00.000")).check();
            checkQuery("SELECT timestamp '1992-01-18 02:30:00' + interval (23) hours")
                .returns(sqlTimestamp("1992-01-19 01:30:00.000")).check();
            checkQuery("SELECT timestamp '1992-01-18 02:30:00' + interval (24) hours")
                .returns(sqlTimestamp("1992-01-19 02:30:00.000")).check();

            // DST started 1992-03-29 02:00:00 -> 1992-03-29 03:00:00
            checkQuery("select date '1992-03-29'").returns(sqlDate("1992-03-29")).check();
            checkQuery("select date '1992-03-28' + interval (1) days").returns(sqlDate("1992-03-29")).check();
            checkQuery("select date '1992-03-28' + interval (24) hours").returns(sqlDate("1992-03-29")).check();
            checkQuery("SELECT timestamp '1992-03-28 02:30:00' + interval (25) hours")
                .returns(sqlTimestamp("1992-03-29 03:30:00.000")).check();
            checkQuery("SELECT timestamp '1992-03-28 02:30:00' + interval (23) hours")
                .returns(sqlTimestamp("1992-03-29 01:30:00.000")).check();
            checkQuery("SELECT timestamp '1992-03-28 02:30:00' + interval (24) hours")
                .returns(sqlTimestamp("1992-03-29 02:30:00.000")).check();

            // DST ended 1992-09-27 03:00:00 -> 1992-09-27 02:00:00
            checkQuery("select date '1992-09-27'").returns(sqlDate("1992-09-27")).check();
            checkQuery("select date '1992-09-26' + interval (1) days").returns(sqlDate("1992-09-27")).check();
            checkQuery("select date '1992-09-26' + interval (24) hours").returns(sqlDate("1992-09-27")).check();
            checkQuery("SELECT timestamp '1992-09-26 02:30:00' + interval (25) hours")
                .returns(sqlTimestamp("1992-09-27 03:30:00.000")).check();
            checkQuery("SELECT timestamp '1992-09-26 02:30:00' + interval (23) hours")
                .returns(sqlTimestamp("1992-09-27 01:30:00.000")).check();
            checkQuery("SELECT timestamp '1992-09-26 02:30:00' + interval (24) hours")
                .returns(sqlTimestamp("1992-09-27 02:30:00.000")).check();

            TimeZone.setDefault(TimeZone.getTimeZone("America/Los_Angeles"));

            // DST ended 2021-11-07 02:00:00 -> 2021-11-07 01:00:00
            checkQuery("select date '2021-11-07'").returns(sqlDate("2021-11-07")).check();
            checkQuery("select date '2021-11-06' + interval (1) days").returns(sqlDate("2021-11-07")).check();
            checkQuery("select date '2021-11-06' + interval (24) hours").returns(sqlDate("2021-11-07")).check();
            checkQuery("SELECT timestamp '2021-11-06 01:30:00' + interval (25) hours")
                .returns(sqlTimestamp("2021-11-07 02:30:00.000")).check();
            // Check string representation here, since after timestamp calculation we have '2021-11-07T01:30:00.000-0800'
            // but Timestamp.valueOf method converts '2021-11-07 01:30:00' in 'America/Los_Angeles' time zone to
            // '2021-11-07T01:30:00.000-0700' (we pass through '2021-11-07 01:30:00' twice after DST ended).
            checkQuery("SELECT (timestamp '2021-11-06 02:30:00' + interval (23) hours)::varchar")
                .returns("2021-11-07 01:30:00").check();
            checkQuery("SELECT (timestamp '2021-11-06 01:30:00' + interval (24) hours)::varchar")
                .returns("2021-11-07 01:30:00").check();
        }
        finally {
            TimeZone.setDefault(oldTz);
        }
    }

    /** */
    public static class DateTimeEntry {
        /** */
        long id;

        /** */
        Date javaDate;

        /** */
        java.sql.Date sqlDate;

        /** */
        Time sqlTime;

        /** */
        Timestamp sqlTimestamp;

        /** */
        public DateTimeEntry(long id, Date javaDate, java.sql.Date sqlDate, Time sqlTime, Timestamp sqlTimestamp) {
            this.id = id;
            this.javaDate = javaDate;
            this.sqlDate = sqlDate;
            this.sqlTime = sqlTime;
            this.sqlTimestamp = sqlTimestamp;
        }
    }

    /** */
    private QueryChecker checkQuery(String qry) {
        return new QueryChecker(qry) {
            @Override protected QueryEngine getEngine() {
                return queryEngine;
            }
        };
    }

    /** */
    private Date javaDate(String str) throws Exception {
        return DATE_FORMAT.parse(str);
    }

    /** */
    private java.sql.Date sqlDate(String str) throws Exception {
        return java.sql.Date.valueOf(str);
    }

    /** */
    private Time sqlTime(String str) throws Exception {
        return Time.valueOf(str);
    }

    /** */
    private Timestamp sqlTimestamp(String str) throws Exception {
        return Timestamp.valueOf(str);
    }
}
